### PR TITLE
Fixes the image feature extractor in observations

### PR DESCRIPTION
**Pull request by Dhoeller19**
# Description

- Fixes the image feature extractor in observations
- Adds missing dependencies in setup.py

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

**Comment by glvov-bdai** [comment]
Hi, this is fixed by this earlier open PR: https://github.com/isaac-sim/IsaacLab/pull/1323

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py [hunk 1]
```diff
@@ -347,61 +375,137 @@ def __call__(
...
+        model_device: str | None = None,
     ) -> torch.Tensor:
-        """Extracted image features from a pre-trained frozen encoder.
+        # obtain the images from the sensor
+        image_data = image(
+            env=env,
+            sensor_cfg=sensor_cfg,
+            data_type=data_type,
+            convert_perspective_to_orthogonal=convert_perspective_to_orthogonal,
+            normalize=True,  # need this for training stability
```
Comment (glvov-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: `normalize=False, # allow user to normalize as part of model preprocess fn `

Comment (glvov-bdai) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: https://github.com/isaac-sim/IsaacLab/pull/1323/files

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py [hunk 2]
```diff
@@ -347,61 +375,137 @@ def __call__(
...
+            """
+            # move the image to the model device
+            img = img.to(model_device)
+            # permute the image to (num_envs, channel, height, width)
+            img = img.permute(0, 3, 1, 2)
+            # normalize the image
+            mean = torch.tensor([0.485, 0.456, 0.406], device=img.device).view(1, 3, 1, 1)
+            std = torch.tensor([0.229, 0.224, 0.225], device=img.device).view(1, 3, 1, 1)
+
+            return (img - mean) / std
```
Comment (glvov-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: "preprocess": (
                    lambda img: (
                        (
                            img.permute(
                                0, 3, 1, 2
                            ).float()  # Convert [batch, height, width, 3] -> [batch, 3, height, width]
                            - torch.tensor([0.485, 0.456, 0.406], device=img.device).view(1, 3, 1, 1)
                        )
                        / torch.tensor([0.229, 0.224, 0.225], device=img.device).view(1, 3, 1, 1)
                    )
                    / 255 # need to normalize this way
                ),

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py [hunk 3]
```diff
@@ -347,61 +375,137 @@ def __call__(
...
+            """Inference the Theia transformer model.
+
+            Args:
+                model: The Theia transformer model.
+                images: The preprocessed image tensor. Shape is (num_envs, height, width, channel).
+
+            Returns:
+                The extracted features tensor. Shape is (num_envs, feature_dim).
+            """
+            return model.forward_feature(images, do_rescale=False, interpolate_pos_encoding=True)
```
Comment (glvov-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: This converts from GPU to CPU when it's not needed: https://github.com/isaac-sim/IsaacLab/pull/1323

                "preprocess": lambda img: {
                    "pixel_values": (  # This deconstructs default inferencing (CPU in transformers) into GPU only operations
                        (  # as advised by first author of Theia paper
                            # Shifts mean, standard deviation, and format to match training set
                            img.permute(
                                0, 3, 1, 2
                            ).float()  # Convert [batch, height, width, 3] -> [batch, 3, height, width]
                            - torch.tensor([0.485, 0.456, 0.406], device=img.device).view(1, 3, 1, 1)
                        )
                        / torch.tensor([0.229, 0.224, 0.225], device=img.device).view(1, 3, 1, 1)
                    ) / 255
                },
                "inference": lambda model, images: model.backbone.model(  # Taken from Transformers; inference converted to be GPU only
                    **images, interpolate_pos_encoding=True
                ).last_hidden_state[
                    :, 1:
                ],

**Review by glvov-bdai — CHANGES_REQUESTED** [review]
Overall LGTM sorry about the device hiccup. I think this PR should include the normalization fixes from https://github.com/isaac-sim/IsaacLab/pull/1323 or https://github.com/isaac-sim/IsaacLab/pull/1323 should be merged first. It also addresses the device issue

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py (hunk 1 above)
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]:
```suggestion
            normalize=False,  # we pre-process based on model
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py (hunk 3 above)
Comment (glvov-bdai) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: Also includes the divide by 255 fix to normalize between 0 and 1, other model won't inference correcly

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py
Comment (jsmith-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py:1]: For fix PRs, it would be good to only fix the broken bits as it's much easier to review than when there a lot of unrelated changes

**Review by jsmith-bdai — APPROVED** [review]
LGTM, thanks for fixing @Mayankm96 @Dhoeller19!

Sorry again for not noticing the test breakage

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py [hunk 4]
```diff
@@ -279,65 +283,89 @@ def image(
 class image_features(ManagerTermBase):
     """Extracted image features from a pre-trained frozen encoder.
 
-    This method calls the :meth:`image` function to retrieve images, and then performs
-    inference on those images.
+    This term uses models from the model zoo in PyTorch and extracts features from the images.
```
Comment (glvov-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py:286]: Also uses models from Transformers

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py [hunk 5]
```diff
@@ -279,65 +283,89 @@ def image(
...
+            else:
+                raise ValueError(
+                    f"Model name '{self.model_name}' not found in the default model zoo configuration."
+                    f" Available models: {default_theia_models + default_resnet_models}."
+                )
+        else:
+            model_config = self.model_zoo_cfg[self.model_name]
+
+        # Retrieve the model, preprocess and inference functions
+        self._model = model_config["model"]()
```
Comment (glvov-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py:370]: By keeping pre-process separate, I think this leaves room for using hydra to override the pre-process to play around with different pre-processing while keeping the model the same. For example we could provide an idempotent pre-process that only divides by 255 instead of normalizing to training set for experimentation

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py [hunk 6]
```diff
@@ -347,61 +375,112 @@ def __call__(
         convert_perspective_to_orthogonal: bool = False,
         model_zoo_cfg: dict | None = None,
         model_name: str = "ResNet18",
-        model_device: str | None = "cuda:0",
-        reset_model: bool = False,
+        model_device: str | None = None,
```
Comment (glvov-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py:389]: Doesn't seem like model device is used here anymore (instead in constructor). Does it still need to be here?

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py [hunk 7]
```diff
@@ -279,65 +283,98 @@ def image(
...
+    The model zoo configuration should be a dictionary that maps different model names to a dictionary
+    that defines the model, preprocess and inference functions. The dictionary should have the following
+    entries:
+
+    - "model": A callable that returns the model when invoked without arguments.
+    - "reset": A callable that resets the model. This is useful when the model has a state that needs to be reset.
+    - "inference": A callable that, when given the model and the images, returns the extracted features.
+
+    If the model zoo configuration is not provided, the default model zoo configurations are used. The default
+    model zoo configurations include the models from Theia and ResNet. These models are loaded from
```
Comment (glvov-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: Maybe link Theia here too? https://theia.theaiinstitute.com/

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py (hunk 5 above)
Comment (glvov-bdai) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py:370]: kwargs could work as per convo with @Mayankm96

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py [hunk 8]
```diff
@@ -347,61 +384,112 @@ def __call__(
...
+
+        Returns:
+            A dictionary containing the model and inference functions.
+        """
+        from torchvision import models
 
-        return features.to(image_device).clone()
+        def _load_model() -> torch.nn.Module:
+            """Load the ResNet model."""
+            model = getattr(models, model_name)(weights="ResNet50_Weights.DEFAULT").eval()
```
Comment (glvov-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: Should support other weights for other sizes as well like  18, 34, 101, etc

along lines of: 


```def create_resnet_model(resnet_name, weights):
    return {
        "model": lambda: getattr(models, resnet_name)(weights=weights).eval().to("cuda:0"),
        "preprocess": (
            lambda img: (
                (
                    img.permute(
                        0, 3, 1, 2
                    ).float()  # Convert [batch, height, width, 3] -> [batch, 3, height, width]
                    - torch.tensor([0.485, 0.456, 0.406], device=img.device).view(1, 3, 1, 1)
                )
                / torch.tensor([0.229, 0.224, 0.225], device=img.device).view(1, 3, 1, 1)
            )
            / 255
        ),
        "inference": lambda model, images: model(images),
    }

resnet_weights = {
    "resnet18": ResNet18_Weights.IMAGENET1K_V1,
    "resnet34": ResNet34_Weights.IMAGENET1K_V1,
    "resnet50": ResNet50_Weights.IMAGENET1K_V1,
    "resnet101": ResNet101_Weights.IMAGENET1K_V1
}

resnet_models = ["resnet18", "resnet34", "resnet50", "resnet101"]
for resnet_name in resnet_models:
    self.default_model_zoo_cfg[resnet_name] = create_resnet_model(resnet_name, resnet_weights[resnet_name])```

**Review by glvov-bdai — APPROVED** [review]
Overall looks good would just update to support more resnet model sizes than just ResNet50

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py (hunk 6 above)
Comment (Mayankm96) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py:389]: yes, we need this still for arg checking...

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py (hunk 4 above)
Comment (Mayankm96) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py:286]: The models are in PyTorch still so this holds. Added references to later in the docs.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py (hunk 6 above)
Comment (Mayankm96) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py:389]: Yes. Still needed for params cross checking.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py (hunk 7 above)
Comment (Mayankm96) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: Added citations to the papers.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py [hunk 9]
```diff
@@ -279,65 +283,98 @@ def image(
...
-                    - torch.tensor([0.485, 0.456, 0.406], device=img.device).view(1, 3, 1, 1)
-                ) / torch.tensor([0.229, 0.224, 0.225], device=img.device).view(1, 3, 1, 1),
-                "inference": lambda model, images: model(images),
-            }
-
-        # List of Theia models
-        theia_models = [
+        # extract parameters from the configuration
+        self.model_zoo_cfg: dict = cfg.params.get("model_zoo_cfg")  # type: ignore
+        self.model_name: str = cfg.params.get("model_name", "ResNet18")  # type: ignore
```
Comment (glvov-bdai) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: Case of this doesn't match the case of resnet list (ResNet18 vs. resnet)

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py (hunk 5 above)
Comment (Mayankm96) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py:370]: Done

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py (hunk 9 above)
Comment (Mayankm96) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/observations.py]: You're looking at an outdated version :)

**Comment by glvov-bdai** [comment]
Overall looks good, I am able to train resnet and theia. 

One note; it may be good to change the feature extraction yaml for RL games cartpole to use 512 or 1024 envs by default, as well as to have a max epoch length of 200 instead of 5k. Can also take care of this in a separate PR but I find these defaults lead to better convergence and less overcomputation respectively; currently the default env count is a little low with a max epoch length of 5k

**Comment by glvov-bdai** [comment]
LGTM! 👈 🤠 👉

**Review by jtigue-bdai — APPROVED** [review]

**Comment by glvov-bdai** [comment]
I suggest  that we delete the ground plane for the cartpole environments. It's not needed, and it leads to behavior where it seems like the feature extraction based cartpole doesn't work when you open up the viewport. Feature extraction cartpole works where the ground plane is not visible. Alternatively, we could e
[Screencast from 10-30-2024 04:37:37 PM.webm](https://github.com/user-attachments/assets/c83e6e1a-fbca-48a7-bad5-3314ddba56be)
xtend the ground plane to cover the whole field

**Comment by Mayankm96** [comment]
@glvov-bdai Removed the ground now

**Comment by Mayankm96** [comment]
An untouched test fails but it passes locally @Dhoeller19 

```
./isaaclab.sh -p source/extensions/omni.isaac.lab/test/utils/test_noise.py
```
